src/schema.c

```diff
@@ -113,14 +135,84 @@ ldap_schema_object_classes(const ldap_schema_t *schema)
 LDAPAttributeType**
 ldap_schema_attribute_types(const ldap_schema_t* schema)
 {
-    if (!schema)
+    if (!schema || !schema->attribute_types_by_oid)
+    {
+        ld_error("ldap_schema_attribute_types - schema or attribute types list is NULL!\n");
+
+        return NULL;
+    }
+
+    int result_size = g_hash_table_size(schema->attribute_types_by_oid);
+
+    LDAPAttributeType** result = talloc_array(schema, LDAPAttributeType*, result_size + 1);
+
+    if (!result)
+    {
+        ld_error("ldap_schema_attribute_types - talloc_array for LDAPAttributeType returned NULL!\n");
+
+        return NULL;
+    }
+
+    GHashTableIter iter;
+    gpointer key = NULL, value = NULL;
+
+    int index = 0;
+    g_hash_table_iter_init(&iter, schema->attribute_types_by_oid);
+    while (g_hash_table_iter_next(&iter, &key, &value)) {
+        result[index] = value;
+        index++;
+    }
+    result[result_size] = NULL;
+
+    return result;
+}
+
+LDAPObjectClass *ldap_schema_get_objectclass_by_oid(const ldap_schema_t* schema, const char *oid)
+{
+    if (!schema || !schema->object_classes_by_oid)
+    {
+        ld_error("ldap_schema_get_objectclass - schema or object classes list is NULL!\n");
+
+        return NULL;
+    }
+
+    return (LDAPObjectClass *)g_hash_table_lookup(schema->object_classes_by_oid, oid);
+}
+
+LDAPObjectClass *ldap_schema_get_objectclass_by_name(const ldap_schema_t* schema, const char *name)
+{
+    if (!schema || !schema->object_classes_by_name)
+    {
+        ld_error("ldap_schema_get_objectclass - schema or object classes list is NULL!\n");
+
+        return NULL;
+    }
+
+    return (LDAPObjectClass *)g_hash_table_lookup(schema->object_classes_by_name, name);
+}
+
+LDAPAttributeType *ldap_schema_get_attributetype_by_oid(const ldap_schema_t* schema, const char *oid)
+{
+    if (!schema || !schema->attribute_types_by_oid)
```

Comment:
Consider separating checks of `schema` and `attribute_types_by_oid`. If we have separate checks for each case, we can accurately say what exactly the error is.